h/api/nipsa/search.py

```diff
@@ -0,0 +1,83 @@
+import copy
+
+
+def nipsa_filter(query, userid=None):
```

Comment:
It's about making this function simple. There's no reason I can see why it needs to accept an argument that it has a contractual obligation not to modify, and therefore has to deep copy, when it never inspects that argument or modifies it anyway, just to prevent the caller from having to know about how to construct filtered queries even though it already knows its using elasticsearch and constructing elasticsearch queries.

The result is only going to be a proliferation of code that modifies elasticsearch query objects, like this code, rather than code that constructs the useful pieces of them to be assembled as needed by the caller.